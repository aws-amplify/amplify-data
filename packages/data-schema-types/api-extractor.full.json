{
  "extends": "../../api-extractor.full.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/index.internal.d.ts"
}
