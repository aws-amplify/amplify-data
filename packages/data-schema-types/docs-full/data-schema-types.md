<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema-types](./data-schema-types.md)

## data-schema-types package

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[DerivedApiDefinition](./data-schema-types.derivedapidefinition.md)


</td><td>


</td></tr>
<tr><td>

[GraphQLFormattedError](./data-schema-types.graphqlformattederror.md)


</td><td>

See: https://spec.graphql.org/draft/\#sec-Errors


</td></tr>
<tr><td>

[SourceLocation](./data-schema-types.sourcelocation.md)


</td><td>

Represents a location in a Source.


</td></tr>
</tbody></table>

## Variables

<table><thead><tr><th>

Variable


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[\_\_modelMeta\_\_](./data-schema-types.__modelmeta__.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[AuthMode](./data-schema-types.authmode.md)


</td><td>


</td></tr>
<tr><td>

[BackendSecret](./data-schema-types.backendsecret.md)


</td><td>

Importing the full objects from @<!-- -->aws-amplify/plugin-types more than doubles dev env runtime. This type replacement will contain the content for config without the negative side-effects. We may need to re-approach if customers interact with these programmatically to avoid forcing narrowing.


</td></tr>
<tr><td>

[BooleanFilters](./data-schema-types.booleanfilters.md)


</td><td>


</td></tr>
<tr><td>

[ContextType](./data-schema-types.contexttype.md)


</td><td>


</td></tr>
<tr><td>

[CreateModelInput](./data-schema-types.createmodelinput.md)


</td><td>

All identifiers and fields used to create a model


</td></tr>
<tr><td>

[CustomHeaders](./data-schema-types.customheaders.md)


</td><td>

Custom headers that can be passed either to the client or to individual model operations, either as a static object or a function that returns a promise.


</td></tr>
<tr><td>

[CustomMutations](./data-schema-types.custommutations.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperationFnParams](./data-schema-types.customoperationfnparams.md)


</td><td>

Generates Custom Operations function params based on whether .arguments() were specified in the schema builder


</td></tr>
<tr><td>

[CustomOperationMethodOptions](./data-schema-types.customoperationmethodoptions.md)


</td><td>


</td></tr>
<tr><td>

[CustomOperations](./data-schema-types.customoperations.md)


</td><td>


</td></tr>
<tr><td>

[CustomQueries](./data-schema-types.customqueries.md)


</td><td>


</td></tr>
<tr><td>

[CustomSelectionSetReturnValue](./data-schema-types.customselectionsetreturnvalue.md)


</td><td>

Generates flattened, readonly return type using specified custom sel. set


</td></tr>
<tr><td>

[CustomSqlDataSourceStrategy](./data-schema-types.customsqldatasourcestrategy.md)


</td><td>


</td></tr>
<tr><td>

[CustomSubscriptions](./data-schema-types.customsubscriptions.md)


</td><td>


</td></tr>
<tr><td>

[DataSourceConfiguration](./data-schema-types.datasourceconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[DatasourceEngine](./data-schema-types.datasourceengine.md)


</td><td>


</td></tr>
<tr><td>

[Debug](./data-schema-types.debug.md)


</td><td>


</td></tr>
<tr><td>

[DeepPickFromPath](./data-schema-types.deeppickfrompath.md)


</td><td>

Picks object properties that match provided dot-separated Path


</td></tr>
<tr><td>

[DeepReadOnlyObject](./data-schema-types.deepreadonlyobject.md)


</td><td>


</td></tr>
<tr><td>

[DefineFunction](./data-schema-types.definefunction.md)


</td><td>


</td></tr>
<tr><td>

[DerivedCombinedSchema](./data-schema-types.derivedcombinedschema.md)


</td><td>


</td></tr>
<tr><td>

[DerivedModelSchema](./data-schema-types.derivedmodelschema.md)


</td><td>


</td></tr>
<tr><td>

[EnumTypes](./data-schema-types.enumtypes.md)


</td><td>

The utility type that is used to infer the type (interface) of the generated `client.enums` property.


</td></tr>
<tr><td>

[Equal](./data-schema-types.equal.md)


</td><td>


</td></tr>
<tr><td>

[ExcludeEmpty](./data-schema-types.excludeempty.md)


</td><td>


</td></tr>
<tr><td>

[Expect](./data-schema-types.expect.md)


</td><td>


</td></tr>
<tr><td>

[ExpectFalse](./data-schema-types.expectfalse.md)


</td><td>


</td></tr>
<tr><td>

[ExpectTrue](./data-schema-types.expecttrue.md)


</td><td>


</td></tr>
<tr><td>

[ExtractModelMeta](./data-schema-types.extractmodelmeta.md)


</td><td>


</td></tr>
<tr><td>

[FunctionSchemaAccess](./data-schema-types.functionschemaaccess.md)


</td><td>


</td></tr>
<tr><td>

[HasKey](./data-schema-types.haskey.md)


</td><td>


</td></tr>
<tr><td>

[IfEquals](./data-schema-types.ifequals.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodsFromIR](./data-schema-types.indexquerymethodsfromir.md)


</td><td>


</td></tr>
<tr><td>

[IndexQueryMethodSignature](./data-schema-types.indexquerymethodsignature.md)


</td><td>


</td></tr>
<tr><td>

[IsAny](./data-schema-types.isany.md)


</td><td>


</td></tr>
<tr><td>

[IsEmptyStringOrNever](./data-schema-types.isemptystringornever.md)


</td><td>


</td></tr>
<tr><td>

[IsFalse](./data-schema-types.isfalse.md)


</td><td>


</td></tr>
<tr><td>

[IsTrue](./data-schema-types.istrue.md)


</td><td>


</td></tr>
<tr><td>

[JsResolver](./data-schema-types.jsresolver.md)


</td><td>


</td></tr>
<tr><td>

[JsResolverEntry](./data-schema-types.jsresolverentry.md)


</td><td>


</td></tr>
<tr><td>

[LambdaFunctionDefinition](./data-schema-types.lambdafunctiondefinition.md)


</td><td>


</td></tr>
<tr><td>

[LazyLoader](./data-schema-types.lazyloader.md)


</td><td>


</td></tr>
<tr><td>

[ListReturnValue](./data-schema-types.listreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[LogicalFilters](./data-schema-types.logicalfilters.md)


</td><td>


</td></tr>
<tr><td>

[Model](./data-schema-types.model.md)


</td><td>


</td></tr>
<tr><td>

[ModelFilter](./data-schema-types.modelfilter.md)


</td><td>


</td></tr>
<tr><td>

[ModelIdentifier](./data-schema-types.modelidentifier.md)


</td><td>


</td></tr>
<tr><td>

[ModelMetaShape](./data-schema-types.modelmetashape.md)


</td><td>


</td></tr>
<tr><td>

[ModelPath](./data-schema-types.modelpath.md)


</td><td>

Generates custom selection set type with up to 6 levels of nested fields


</td></tr>
<tr><td>

[ModelSortDirection](./data-schema-types.modelsortdirection.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypes](./data-schema-types.modeltypes.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesClient](./data-schema-types.modeltypesclient.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRCookies](./data-schema-types.modeltypesssrcookies.md)


</td><td>


</td></tr>
<tr><td>

[ModelTypesSSRRequest](./data-schema-types.modeltypesssrrequest.md)


</td><td>


</td></tr>
<tr><td>

[MutationInput](./data-schema-types.mutationinput.md)


</td><td>

All required fields and relational fields, exclude readonly fields


</td></tr>
<tr><td>

[NonRelationalFields](./data-schema-types.nonrelationalfields.md)


</td><td>

Currently this omits any object-type fields. Update this when we add custom types/enums.


</td></tr>
<tr><td>

[NotAny](./data-schema-types.notany.md)


</td><td>


</td></tr>
<tr><td>

[NotEqual](./data-schema-types.notequal.md)


</td><td>


</td></tr>
<tr><td>

[NumericFilter](./data-schema-types.numericfilter.md)


</td><td>


</td></tr>
<tr><td>

[ObjectIsNonEmpty](./data-schema-types.objectisnonempty.md)


</td><td>


</td></tr>
<tr><td>

[ObservedReturnValue](./data-schema-types.observedreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[ObserveQueryReturnValue](./data-schema-types.observequeryreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[PathEntry](./data-schema-types.pathentry.md)


</td><td>

references IAmplifyGraphqlDefinition from: https://github.com/aws-amplify/amplify-category-api/blob/4c0ea253a0bae51f775383929ba4748593185bc1/packages/amplify-graphql-api-construct/src/types.ts\#L491-L503

function slots is any'd for now. Will add actual type when we add support for this feature


</td></tr>
<tr><td>

[Prettify](./data-schema-types.prettify.md)


</td><td>


</td></tr>
<tr><td>

[RequestOptions](./data-schema-types.requestoptions.md)


</td><td>

Request options that are passed to custom header functions. `method` and `headers` are not included in custom header functions passed to subscriptions.


</td></tr>
<tr><td>

[ResolvedModel](./data-schema-types.resolvedmodel.md)


</td><td>

Flattens model instance type and unwraps async functions into resolved GraphQL shape

This type is used for generating the base shape for custom selection set input and its return value Uses same pattern as above to limit recursion depth to maximum usable for selection set.


</td></tr>
<tr><td>

[RestoreArrays](./data-schema-types.restorearrays.md)


</td><td>

This mapped type traverses the SelectionSetReturnValue result and the original FlatModel, restoring array types that were flattened in DeepPickFromPath


</td></tr>
<tr><td>

[ReturnValue](./data-schema-types.returnvalue.md)


</td><td>

Selection set-aware CRUDL operation return value type


</td></tr>
<tr><td>

[SchemaConfiguration](./data-schema-types.schemaconfiguration.md)


</td><td>


</td></tr>
<tr><td>

[SecondaryIndexIrShape](./data-schema-types.secondaryindexirshape.md)


</td><td>

SecondaryIndex index types and query methods


</td></tr>
<tr><td>

[SelectionSet](./data-schema-types.selectionset.md)


</td><td>


</td></tr>
<tr><td>

[SetTypeSubArg](./data-schema-types.settypesubarg.md)


</td><td>

Replaces the value of a key in a complex generic type param


</td></tr>
<tr><td>

[SingularReturnValue](./data-schema-types.singularreturnvalue.md)


</td><td>


</td></tr>
<tr><td>

[SizeFilter](./data-schema-types.sizefilter.md)


</td><td>

Filter options that can be used on fields where size checks are supported.


</td></tr>
<tr><td>

[SqlStatementFolderEntry](./data-schema-types.sqlstatementfolderentry.md)


</td><td>


</td></tr>
<tr><td>

[StringFilter](./data-schema-types.stringfilter.md)


</td><td>

Filters options that can be used on string-like fields.


</td></tr>
<tr><td>

[SubnetAZ](./data-schema-types.subnetaz.md)


</td><td>


</td></tr>
<tr><td>

[UnionToIntersection](./data-schema-types.uniontointersection.md)


</td><td>


</td></tr>
<tr><td>

[UnwrapArray](./data-schema-types.unwraparray.md)


</td><td>


</td></tr>
<tr><td>

[VpcConfig](./data-schema-types.vpcconfig.md)


</td><td>


</td></tr>
<tr><td>

[WritableKeys](./data-schema-types.writablekeys.md)


</td><td>


</td></tr>
</tbody></table>
