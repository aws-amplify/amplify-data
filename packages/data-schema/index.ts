import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
