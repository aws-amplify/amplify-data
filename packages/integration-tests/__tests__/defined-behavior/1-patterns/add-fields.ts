import { a, ClientSchema } from '@aws-amplify/data-schema';
import { Amplify } from 'aws-amplify';
import { Expect, Equal } from '@aws-amplify/data-schema-types';
import {
  buildAmplifyConfig,
  mockedGenerateClient,
  optionsAndHeaders,
} from '../../utils';

describe('Specify a custom field type', () => {
  // https://docs.amplify.aws/react/build-a-backend/data/data-modeling/add-fields/#specify-a-custom-field-type

  // #region test case data
  const samplePost = {
    __typename: 'Post',
    id: 'some-id',
    content: 'some content',
    location: { lat: 1, long: 2 },
    updatedAt: '2024-03-01T19:05:44.536Z',
    createdAt: '2024-03-01T18:05:44.536Z',
  } as const;

  const opInputResultMap = {
    create: {
      input: {
        content: 'hello',
        location: { lat: 3, long: 4 },
      },
      resultKey: 'createPost',
      result: {
        ...samplePost,
        content: 'hello',
        location: { lat: 3, long: 4 },
      },
    },
    update: {
      input: {
        id: samplePost.id,
        location: { lat: 5, long: 6 },
      },
      resultKey: 'updatePost',
      result: {
        ...samplePost,
        location: { lat: 5, long: 6 },
      },
    },
    delete: {
      input: {
        id: samplePost.id,
      },
      resultKey: 'deletePost',
      result: samplePost,
    },
    list: {
      input: undefined,
      resultKey: 'listPosts',
      result: { items: [samplePost] },
    },
    get: {
      input: {
        id: samplePost.id,
      },
      resultKey: 'getPost',
      result: samplePost,
    },
  } as const;
  // #endregion

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Inline definition', () => {
    // #region covers 279bf864a21cd97b
    const schema = a
      .schema({
        Post: a.model({
          location: a.customType({
            lat: a.float(),
            long: a.float(),
          }),
          content: a.string(),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    describe('types', () => {
      type ExpectedCustomTypeInput =
        | {
            lat?: number | null;
            long?: number | null;
          }
        | null
        | undefined;

      test('client input types have expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<
            Schema['Post']['createType']['location'],
            ExpectedCustomTypeInput
          >
        >;

        type _ExpectedUpdateInput = Expect<
          Equal<
            Schema['Post']['createType']['location'],
            ExpectedCustomTypeInput
          >
        >;
      });

      test('client return type has expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<Schema['Post']['type']['location'], ExpectedCustomTypeInput>
        >;
      });
    });

    describe('client runtime', () => {
      Object.entries(opInputResultMap).forEach(
        ([op, { input, resultKey, result }]) => {
          test(`${op}`, async () => {
            const { spy, generateClient } = mockedGenerateClient([
              {
                data: { [resultKey]: result },
              },
            ]);
            const config = await buildAmplifyConfig(schema);
            Amplify.configure(config);
            const client = generateClient<Schema>();

            // @ts-ignore - testing runtime behavior; types are not relevant
            const { data, errors } = await client.models.Post[op](input);

            expect(optionsAndHeaders(spy)).toMatchSnapshot();
            expect(errors).toBeUndefined();

            const normalizedResult =
              op === 'list'
                ? (result as (typeof opInputResultMap)['list']['result']).items
                : result;

            expect(data).toEqual(normalizedResult);
          });
        },
      );
    });
  });

  describe('Explicit definition', () => {
    // #region covers b7fd0a0c85d7fa05
    const schema = a
      .schema({
        Location: a.customType({
          lat: a.float(),
          long: a.float(),
        }),

        Post: a.model({
          location: a.ref('Location'),
          content: a.string(),
        }),

        User: a.model({
          lastKnownLocation: a.ref('Location'),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    describe('types', () => {
      type ExpectedCustomTypeInput =
        | {
            lat?: number | null;
            long?: number | null;
          }
        | null
        | undefined;

      test('client input types have expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<
            Schema['Post']['createType']['location'],
            ExpectedCustomTypeInput
          >
        >;

        type _ExpectedUpdateInput = Expect<
          Equal<
            Schema['Post']['createType']['location'],
            ExpectedCustomTypeInput
          >
        >;
      });

      test('client return type has expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<Schema['Post']['type']['location'], ExpectedCustomTypeInput>
        >;
      });
    });

    describe('client runtime', () => {
      Object.entries(opInputResultMap).forEach(
        ([op, { input, resultKey, result }]) => {
          test(`${op}`, async () => {
            const { spy, generateClient } = mockedGenerateClient([
              {
                data: { [resultKey]: result },
              },
            ]);
            const config = await buildAmplifyConfig(schema);
            Amplify.configure(config);
            const client = generateClient<Schema>();

            // @ts-ignore - testing runtime behavior; types are not relevant
            const { data, errors } = await client.models.Post[op](input);

            expect(optionsAndHeaders(spy)).toMatchSnapshot();
            expect(errors).toBeUndefined();

            const normalizedResult =
              op === 'list'
                ? (result as (typeof opInputResultMap)['list']['result']).items
                : result;

            expect(data).toEqual(normalizedResult);
          });
        },
      );
    });
  });
});

describe('Specify an enum field type', () => {
  // https://docs.amplify.aws/react/build-a-backend/data/data-modeling/add-fields/#specify-an-enum-field-type
  afterEach(() => {
    jest.clearAllMocks();
  });

  // #region test data
  const samplePost = {
    __typename: 'Post',
    id: 'some-id',
    content: 'some content',
    privacySetting: 'PUBLIC',
    updatedAt: '2024-03-01T19:05:44.536Z',
    createdAt: '2024-03-01T18:05:44.536Z',
  } as const;

  const opInputResultMap = {
    create: {
      input: {
        content: 'hello',
        privacySetting: 'PRIVATE',
      },
      resultKey: 'createPost',
      result: {
        ...samplePost,
        content: 'hello',
        privacySetting: 'PRIVATE',
      },
    },
    update: {
      input: {
        id: samplePost.id,
        privacySetting: 'FRIENDS_ONLY',
      },
      resultKey: 'updatePost',
      result: {
        ...samplePost,
        privacySetting: 'FRIENDS_ONLY',
      },
    },
    delete: {
      input: {
        id: samplePost.id,
      },
      resultKey: 'deletePost',
      result: samplePost,
    },
    list: {
      input: undefined,
      resultKey: 'listPosts',
      result: { items: [samplePost] },
    },
    get: {
      input: {
        id: samplePost.id,
      },
      resultKey: 'getPost',
      result: samplePost,
    },
  } as const;
  // #endregion

  describe('Inline definition', () => {
    // #region covers 35bf03594722226f
    const schema = a
      .schema({
        Post: a.model({
          privacySetting: a.enum(['PRIVATE', 'FRIENDS_ONLY', 'PUBLIC']),
          content: a.string(),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    describe('types', () => {
      type ExpectedEnumInput =
        | 'PRIVATE'
        | 'FRIENDS_ONLY'
        | 'PUBLIC'
        | null
        | undefined;

      test('client input types have expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<
            Schema['Post']['createType']['privacySetting'],
            ExpectedEnumInput
          >
        >;

        type _ExpectedUpdateInput = Expect<
          Equal<
            Schema['Post']['createType']['privacySetting'],
            ExpectedEnumInput
          >
        >;
      });

      test('client return type has expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<Schema['Post']['type']['privacySetting'], ExpectedEnumInput>
        >;
      });
    });

    describe('client runtime', () => {
      test('list enum values', async () => {
        const { generateClient } = mockedGenerateClient([]);
        const config = await buildAmplifyConfig(schema);
        Amplify.configure(config);
        const client = generateClient<Schema>();

        const enumValues = client.enums.PostPrivacySetting.values();
        expect(enumValues).toStrictEqual(['PRIVATE', 'FRIENDS_ONLY', 'PUBLIC']);
      });

      Object.entries(opInputResultMap).forEach(
        ([op, { input, resultKey, result }]) => {
          test(`${op}`, async () => {
            const { spy, generateClient } = mockedGenerateClient([
              {
                data: { [resultKey]: result },
              },
            ]);
            const config = await buildAmplifyConfig(schema);
            Amplify.configure(config);
            const client = generateClient<Schema>();

            // @ts-ignore - testing runtime behavior; types are not relevant
            const { data, errors } = await client.models.Post[op](input);

            expect(optionsAndHeaders(spy)).toMatchSnapshot();
            expect(errors).toBeUndefined();

            const normalizedResult =
              op === 'list'
                ? (result as (typeof opInputResultMap)['list']['result']).items
                : result;

            expect(data).toEqual(normalizedResult);
          });
        },
      );
    });
  });

  describe('Explicit definition', () => {
    // #region covers f88634036e3d5189
    const schema = a
      .schema({
        PrivacySetting: a.enum(['PRIVATE', 'FRIENDS_ONLY', 'PUBLIC']),

        Post: a.model({
          privacySetting: a.ref('PrivacySetting'),
          content: a.string(),
        }),

        Video: a.model({
          privacySetting: a.ref('PrivacySetting'),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    describe('types', () => {
      type ExpectedEnumInput =
        | 'PRIVATE'
        | 'FRIENDS_ONLY'
        | 'PUBLIC'
        | null
        | undefined;

      test('client input types have expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<
            Schema['Post']['createType']['privacySetting'],
            ExpectedEnumInput
          >
        >;

        type _ExpectedUpdateInput = Expect<
          Equal<
            Schema['Post']['createType']['privacySetting'],
            ExpectedEnumInput
          >
        >;
      });

      test('client return type has expected shape', () => {
        type _ExpectedCreateInput = Expect<
          Equal<Schema['Post']['type']['privacySetting'], ExpectedEnumInput>
        >;
      });
    });

    describe('client runtime', () => {
      test('list enum values', async () => {
        const { generateClient } = mockedGenerateClient([]);
        const config = await buildAmplifyConfig(schema);
        Amplify.configure(config);
        const client = generateClient<Schema>();

        // #region covers 1f73a4fff6b93118
        const enumValues = client.enums.PrivacySetting.values();
        expect(enumValues).toStrictEqual(['PRIVATE', 'FRIENDS_ONLY', 'PUBLIC']);
        // #endregion
      });

      Object.entries(opInputResultMap).forEach(
        ([op, { input, resultKey, result }]) => {
          test(`${op}`, async () => {
            const { spy, generateClient } = mockedGenerateClient([
              {
                data: { [resultKey]: result },
              },
            ]);
            const config = await buildAmplifyConfig(schema);
            Amplify.configure(config);
            const client = generateClient<Schema>();

            // @ts-ignore - testing runtime behavior; types are not relevant
            const { data, errors } = await client.models.Post[op](input);

            expect(optionsAndHeaders(spy)).toMatchSnapshot();
            expect(errors).toBeUndefined();

            const normalizedResult =
              op === 'list'
                ? (result as (typeof opInputResultMap)['list']['result']).items
                : result;

            expect(data).toEqual(normalizedResult);
          });
        },
      );
    });
  });

  describe('Mark field as required', () => {
    // #region covers 0d26c55b7f416673
    const schema = a
      .schema({
        Todo: a.model({
          content: a.string().required(),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    test('Required field is non-nullable', () => {
      type _test = Expect<Equal<Schema['Todo']['type']['content'], string>>;
    });

    test('Required field must be provided during create', async () => {
      const { generateClient } = mockedGenerateClient([]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();

      // @ts-expect-error
      await expect(client.models.Todo.create({})).rejects.toThrow();
    });
  });

  describe('Mark field as array', () => {
    // #region covers 0d26c55b7f416673
    const schema = a
      .schema({
        Todo: a.model({
          content: a.string().required(),
          todos: a.string().array(),
        }),
      })
      .authorization((allow) => allow.publicApiKey());
    // #endregion

    type Schema = ClientSchema<typeof schema>;

    test('Array field is an array type', () => {
      type _test = Expect<
        Equal<
          Schema['Todo']['type']['todos'],
          (string | null)[] | null | undefined
        >
      >;
    });

    test('Accepts arrays during create', async () => {
      const { generateClient } = mockedGenerateClient([
        {
          data: {
            createTodo: {
              id: 'some-id',
              content: 'some-content',
              todos: ['a', 'b', 'c'],
            },
          },
        },
      ]);
      const config = await buildAmplifyConfig(schema);
      Amplify.configure(config);
      const client = generateClient<Schema>();

      const { data } = await client.models.Todo.create({
        content: 'some-content',
        todos: ['a', 'b', 'c'],
      });

      expect(data).toEqual(
        expect.objectContaining({
          content: 'some-content',
          todos: ['a', 'b', 'c'],
        }),
      );
    });
  });
});
