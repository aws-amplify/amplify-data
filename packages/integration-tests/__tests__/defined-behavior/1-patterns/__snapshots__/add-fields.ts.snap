// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specify a custom field type Explicit definition client runtime create 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreatePostInput!) { createPost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "content": "hello",
          "location": {
            "lat": 3,
            "long": 4,
          },
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Explicit definition client runtime delete 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: DeletePostInput!) { deletePost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Explicit definition client runtime get 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getPost(id: $id) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "id": "some-id",
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Explicit definition client runtime list 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelPostFilterInput,$limit: Int,$nextToken: String) { listPosts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id location { lat long } content createdAt updatedAt } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Explicit definition client runtime update 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdatePostInput!) { updatePost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
          "location": {
            "lat": 5,
            "long": 6,
          },
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Inline definition client runtime create 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreatePostInput!) { createPost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "content": "hello",
          "location": {
            "lat": 3,
            "long": 4,
          },
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Inline definition client runtime delete 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: DeletePostInput!) { deletePost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Inline definition client runtime get 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getPost(id: $id) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "id": "some-id",
      },
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Inline definition client runtime list 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelPostFilterInput,$limit: Int,$nextToken: String) { listPosts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id location { lat long } content createdAt updatedAt } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;

exports[`Specify a custom field type Inline definition client runtime update 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdatePostInput!) { updatePost(input: $input) { id location { lat long } content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
          "location": {
            "lat": 5,
            "long": 6,
          },
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Explicit definition client runtime create 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreatePostInput!) { createPost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "content": "hello",
          "privacySetting": "PRIVATE",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Explicit definition client runtime delete 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: DeletePostInput!) { deletePost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Explicit definition client runtime get 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getPost(id: $id) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "id": "some-id",
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Explicit definition client runtime list 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelPostFilterInput,$limit: Int,$nextToken: String) { listPosts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id privacySetting content createdAt updatedAt } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Explicit definition client runtime update 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdatePostInput!) { updatePost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
          "privacySetting": "FRIENDS_ONLY",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Inline definition client runtime create 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: CreatePostInput!) { createPost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "content": "hello",
          "privacySetting": "PRIVATE",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Inline definition client runtime delete 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: DeletePostInput!) { deletePost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
        },
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Inline definition client runtime get 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($id: ID!) { getPost(id: $id) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "id": "some-id",
      },
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Inline definition client runtime list 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "query($filter: ModelPostFilterInput,$limit: Int,$nextToken: String) { listPosts(filter: $filter,limit: $limit,nextToken: $nextToken) { items { id privacySetting content createdAt updatedAt } nextToken __typename } }",
      "variables": {},
    },
    {},
  ],
]
`;

exports[`Specify an enum field type Inline definition client runtime update 1`] = `
[
  [
    {
      "authMode": undefined,
      "authToken": undefined,
      "query": "mutation($input: UpdatePostInput!) { updatePost(input: $input) { id privacySetting content createdAt updatedAt } }",
      "variables": {
        "input": {
          "id": "some-id",
          "privacySetting": "FRIENDS_ONLY",
        },
      },
    },
    {},
  ],
]
`;
